openedx/core/lib/api/tests/test_parsers.py

```diff
@@ -0,0 +1,141 @@
+        specified extensions
+        """
+        request = self.request_factory.post(
+            '/',
+            content_type='application/octet-stream',
+            HTTP_CONTENT_DISPOSITION='attachment; filename="VIRUS.EXE',
+        )
+        context = {'view': self.view, 'request': request}
+        result = self.parser.parse(
+            stream=BytesIO('abcdefgh'), media_type='application/octet-stream', parser_context=context
+        )
+        self.assertEqual(result.data, {})
+        self.assertIn('file', result.files)
+        self.assertEqual(result.files['file'].read(), 'abcdefgh')
+
+    def test_parse_mismatched_filename_and_mimetype(self):
+        """
+        Test that TypedFileUploadParser raises an exception when the specified
+        content-type doesn't match the filename extension in the
+        content-disposition header.
+        """
+        request = self.request_factory.post(
+            '/',
+            content_type='image/png',
+            HTTP_CONTENT_DISPOSITION='attachment; filename="file.jpg"',
+        )
+        context = {'view': self.view, 'request': request}
+        with self.assertRaises(exceptions.ParseError) as err:
+            self.parser.parse(stream=BytesIO('abcdefgh'), media_type='image/png', parser_context=context)
+            self.assertIn('user_message', err.detail)
+            self.assertIn('developer_message', err.detail)
+
+    def test_parse_any_type(self):
+        """
+        The view does not have to specify supported types.
+        """
+        view = object()
+        self.assertFalse(hasattr(view, 'upload_media_types'))
+
+        request = self.request_factory.post(
+            '/',
+            content_type='image/png',
+            HTTP_CONTENT_DISPOSITION='attachment; filename="file.png"',
+        )
+        context = {'view': view, 'request': request}
+        result = self.parser.parse(stream=BytesIO('abcdefgh'), media_type='image/png', parser_context=context)
+        self.assertEqual(result.data, {})
+        self.assertIn('file', result.files)
+        self.assertEqual(result.files['file'].read(), 'abcdefgh')
+
+    def test_parse_any_type_with_mismatched_extension(self):
+        """
+        When the view doesn't specify a set of supported types, the
+        TypedFileUploadParser will still check that the extension
+        and content-type match.
+        """
+        view = object()
+        self.assertFalse(hasattr(view, 'upload_media_types'))
+        request = self.request_factory.post(
+            '/',
+            content_type='image/png',
+            HTTP_CONTENT_DISPOSITION='attachment; filename="file.jpg"',
+        )
+        context = {'view': view, 'request': request}
+        with self.assertRaises(exceptions.ParseError):
+            self.parser.parse(stream=BytesIO('abcdefgh'), media_type='image/png', parser_context=context)
+
+    def test_parse_all_types_with_unknown_type(self):
+        """
+        No match checking is done if
```

Comment:
I hear that.  Here's my rationale for the current behavior:

1. The Parser's list of type-to-extension mappings should be canonical.  Right now it only handles image types, because that's the use case for this ticket.
2. The view is responsible for constraining the accepted types.  If the view specifies upload_media_types, it only accepts those types.  If someone wanted to add a View that used this for (say) audio files, I'd hope they would add the mapping for mp3s, oggs, flacs, or whatever they wanted to support, but it should still be added on the parser, and only need to be done once.
3. There may be certain types that don't have a known extension that should allow anything through (like application/octet-stream).  This is the mechanism for supporting those.
4. If the view doesn't constrain accepted mimetypes (though it probably should), the parser will check for matching extensions where possible, but can't necessarily know about everything.
5. We could hook this view in to a comprehensive source of mimetype-to-extension mappings, but there are A LOT of mimetypes out there, most of which we'll probably never want to accept.  I'm not sure it's worth adding that overhead, at least not now.  I chose to start small.  The current use of this parser has constrained types, and knows the extension mappings for all of them.